File: src/api/src/routes/ai.ts
import { Router, Request, Response } from 'express';
import { generateSummary, streamChat, draftCareNote } from '../ai/agent';

const router = Router();

router.post('/summary/:residentId', async (req: Request, res: Response) => {
  try {
    const residentId = parseInt(req.params.residentId, 10);
    const summary = await generateSummary(residentId);
    res.json({ summary });
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

router.post('/chat/:residentId', async (req: Request, res: Response) => {
  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection', 'keep-alive');
  res.flushHeaders();

  const residentId = parseInt(req.params.residentId, 10);
  const { message, history = [] } = req.body as {
    message: string;
    history: Array<{ role: 'user' | 'assistant'; content: string }>;
  };

  try {
    await streamChat(residentId, message, history, (text) => {
      res.write(`data: ${JSON.stringify({ text })}\n\n`);
    });
    res.write('data: [DONE]\n\n');
  } catch (err) {
    res.write(`data: ${JSON.stringify({ error: String(err) })}\n\n`);
  } finally {
    res.end();
  }
});

router.post('/care-note/:residentId', async (req: Request, res: Response) => {
  try {
    const residentId = parseInt(req.params.residentId, 10);
    const { shiftNotes = '' } = req.body as { shiftNotes: string };
    const note = await draftCareNote(residentId, shiftNotes);
    res.json({ note });
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

export default router;

File: src/api-python/routes/ai.py
from __future__ import annotations

import asyncio
import json

from fastapi import APIRouter
from fastapi.responses import StreamingResponse

from ai.agent import draft_care_note, generate_summary, stream_chat_chunks
from ai.context import build_resident_context
from models import CareNoteRequest, ChatRequest

router = APIRouter()


@router.post("/summary/{resident_id}")
async def summary(resident_id: int) -> dict:
    context = await asyncio.to_thread(build_resident_context, resident_id)
    text = await generate_summary(context)
    return {"summary": text}


@router.post("/chat/{resident_id}")
async def chat(resident_id: int, body: ChatRequest) -> StreamingResponse:
    context = await asyncio.to_thread(build_resident_context, resident_id)

    async def generate():
        try:
            async for text in stream_chat_chunks(context, body.message, body.history):
                yield f"data: {json.dumps({'text': text})}\n\n"
            yield "data: [DONE]\n\n"
        except Exception as exc:
            yield f"data: {json.dumps({'error': str(exc)})}\n\n"

    return StreamingResponse(
        generate(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "Connection": "keep-alive"},
    )


@router.post("/care-note/{resident_id}")
async def care_note(resident_id: int, body: CareNoteRequest) -> dict:
    context = await asyncio.to_thread(build_resident_context, resident_id)
    note = await draft_care_note(context, body.shiftNotes)
    return {"note": note}
